File: tablet/Resources/task/Icd10Depressive.js
// Icd10Depressive.js

/*
    Copyright (C) 2012-2015 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true, nomen: true, plusplus: true */
"use strict";
/*global Titanium, L */

var DBCONSTANTS = require('common/DBCONSTANTS'),
    dbcommon = require('lib/dbcommon'),
    taskcommon = require('lib/taskcommon'),
    lang = require('lib/lang'),
    tablename = "icd10depressive",
    fieldlist = dbcommon.standardTaskFields(),
    CORE_NAMES = [
        'mood',
        'anhedonia',
        'energy'
    ],
    ADDITIONAL_NAMES = [
        'sleep',
        'worth',
        'appetite',
        'guilt',
        'concentration',
        'activity',
        'death'
    ],
    SOMATIC_NAMES = [
        'somatic_anhedonia',
        'somatic_emotional_unreactivity',
        'somatic_early_morning_waking',
        'somatic_mood_worse_morning',
        'somatic_psychomotor',
        'somatic_appetite',
        'somatic_weight',
        'somatic_libido'
    ],
    PSYCHOSIS_NAMES = [
        'hallucinations_schizophrenic',
        'hallucinations_other',
        'delusions_schizophrenic',
        'delusions_other',
        'stupor'
    ];

fieldlist.push.apply(fieldlist, dbcommon.CLINICIAN_FIELDSPECS); // Clinician info 1/3
fieldlist.push(
    {name: 'date_pertains_to', type: DBCONSTANTS.TYPE_DATE},
    {name: 'comments', type: DBCONSTANTS.TYPE_TEXT},
    {name: 'duration_at_least_2_weeks', type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: 'severe_clinically', type: DBCONSTANTS.TYPE_BOOLEAN}
);
dbcommon.appendBasicFieldsFromNameArray(fieldlist, CORE_NAMES,
                                        DBCONSTANTS.TYPE_BOOLEAN);
dbcommon.appendBasicFieldsFromNameArray(fieldlist, ADDITIONAL_NAMES,
                                        DBCONSTANTS.TYPE_BOOLEAN);
dbcommon.appendBasicFieldsFromNameArray(fieldlist, SOMATIC_NAMES,
                                        DBCONSTANTS.TYPE_BOOLEAN);
dbcommon.appendBasicFieldsFromNameArray(fieldlist, PSYCHOSIS_NAMES,
                                        DBCONSTANTS.TYPE_BOOLEAN);

// CREATE THE TABLE

dbcommon.createTable(tablename, fieldlist);

// TASK

function Icd10Depressive(patient_id) {
    taskcommon.BaseTask.call(this, patient_id); // call base constructor
    var moment = require('lib/moment');
    // Default values
    this.date_pertains_to = moment(); // Default is today
}

lang.inheritPrototype(Icd10Depressive, taskcommon.BaseTask);
lang.extendPrototype(Icd10Depressive, {

    // KEY DATABASE FIELDS (USED BY DatabaseObject)

    _objecttype: Icd10Depressive,
    _tablename: tablename,
    _fieldlist: fieldlist,

    // TASK CLASS FIELD OVERRIDES (USED BY BaseTask)

    // OTHER

    // Scoring
    n_core: function () {
        return taskcommon.countBooleansByFieldnameArray(this, CORE_NAMES);
    },

    n_additional: function () {
        return taskcommon.countBooleansByFieldnameArray(this, ADDITIONAL_NAMES);
    },

    n_total: function () {
        return this.n_core() + this.n_additional();
    },

    n_somatic: function () {
        return taskcommon.countBooleansByFieldnameArray(this, SOMATIC_NAMES);
    },

    main_complete: function () {
        return (
            this.duration_at_least_2_weeks !== null &&
            taskcommon.isCompleteByFieldnameArray(this, CORE_NAMES) &&
            taskcommon.isCompleteByFieldnameArray(this, ADDITIONAL_NAMES)
        ) || (
            this.severe_clinically
        );
    },

    // Meets criteria? These also return null for unknown.
    meets_criteria_severe_psychotic_schizophrenic: function () {
        var x = this.meets_criteria_severe_ignoring_psychosis();
        if (!x) {
            return x;
        }
        if (this.stupor || this.hallucinations_other || this.delusions_other) {
            return false; // that counts as F32.3
        }
        if (this.stupor === null || this.hallucinations_other === null ||
                this.delusions_other === null) {
            return null; // might be F32.3
        }
        if (this.hallucinations_schizophrenic || this.delusions_schizophrenic) {
            return true;
        }
        if (this.hallucinations_schizophrenic === null ||
                this.delusions_schizophrenic === null) {
            return null;
        }
        return false;
    },

    meets_criteria_severe_psychotic_icd: function () {
        var x = this.meets_criteria_severe_ignoring_psychosis();
        if (!x) {
            return x;
        }
        if (this.stupor || this.hallucinations_other || this.delusions_other) {
            return true;
        }
        if (this.stupor === null || this.hallucinations_other === null ||
                this.delusions_other === null) {
            return null;
        }
        return false;
    },

    meets_criteria_severe_nonpsychotic: function () {
        var x = this.meets_criteria_severe_ignoring_psychosis();
        if (!x) {
            return x;
        }
        if (!taskcommon.isCompleteByFieldnameArray(this, PSYCHOSIS_NAMES)) {
            return null;
        }
        return (
            taskcommon.countBooleansByFieldnameArray(this,
                                                     PSYCHOSIS_NAMES) === 0
        );
    },

    meets_criteria_severe_ignoring_psychosis: function () {
        if (this.severe_clinically) {
            return true;
        }
        if (this.duration_at_least_2_weeks !== null &&
                !this.duration_at_least_2_weeks) {
            return false; // too short
        }
        if (this.n_core() >= 3 && this.n_total() >= 8) {
            return true;
        }
        if (!this.main_complete()) {
            return null; // addition of more information might increase severity
        }
        return false;
    },

    meets_criteria_moderate: function () {
        if (this.severe_clinically) {
            return false; // too severe
        }
        if (this.duration_at_least_2_weeks !== null &&
                !this.duration_at_least_2_weeks) {
            return false; // too short
        }
        if (this.n_core() >= 3 && this.n_total() >= 8) {
            return false; // too severe; that's severe
        }
        if (!this.main_complete()) {
            return null;
            // addition of more information might increase severity
        }
        if (this.n_core() >= 2 && this.n_total() >= 6) {
            return true;
        }
        return false;
    },

    meets_criteria_mild: function () {
        if (this.severe_clinically) {
            return false; // too severe
        }
        if (this.duration_at_least_2_weeks !== null &&
                !this.duration_at_least_2_weeks) {
            return false; // too short
        }
        if (this.n_core() >= 2 && this.n_total() >= 6) {
            return false; // too severe; that's moderate
        }
        if (!this.main_complete()) {
            return null;
            // addition of more information might increase severity
        }
        if (this.n_core() >= 2 && this.n_total() >= 4) {
            return true;
        }
        return false;
    },

    meets_criteria_none: function () {
        if (this.severe_clinically) {
            return false; // too severe
        }
        if (this.duration_at_least_2_weeks !== null &&
                !this.duration_at_least_2_weeks) {
            return true; // too short for depression
        }
        if (this.n_core() >= 2 && this.n_total() >= 4) {
            return false; // too severe
        }
        if (!this.main_complete()) {
            return null;
            // addition of more information might increase severity
        }
        return true;
    },

    meets_criteria_somatic: function () {
        var t = this.n_somatic(),
            u = taskcommon.numIncompleteByFieldnameArray(this, SOMATIC_NAMES);
        if (t >= 4) {
            return true;
        }
        if (t + u < 4) {
            return false;
        }
        return null;
    },

    get_somatic_description: function () {
        var x = this.meets_criteria_somatic();
        if (x === null) {
            return L("icd10depressive_category_somatic_unknown");
        }
        if (x) {
            return L("icd10depressive_category_with_somatic");
        }
        return L("icd10depressive_category_without_somatic");
    },

    get_main_description: function () {
        if (this.meets_criteria_severe_psychotic_schizophrenic()) {
            return L("icd10depressive_category_severe_psychotic_schizophrenic");
        }
        if (this.meets_criteria_severe_psychotic_icd()) {
            return L("icd10depressive_category_severe_psychotic");
        }
        if (this.meets_criteria_severe_nonpsychotic()) {
            return L("icd10depressive_category_severe_nonpsychotic");
        }
        if (this.meets_criteria_moderate()) {
            return L("icd10depressive_category_moderate");
        }
        if (this.meets_criteria_mild()) {
            return L("icd10depressive_category_mild");
        }
        if (this.meets_criteria_none()) {
            return L("icd10depressive_category_none");
        }
        return L("Unknown");
    },

    get_full_description: function () {
        var skipSomatic = this.main_complete() && this.meets_criteria_none();
        return (
            this.get_main_description() +
            (skipSomatic ? "" : " " + this.get_somatic_description())
        );
    },

    // Standard task functions
    isComplete: function () {
        return (
            this.date_pertains_to !== null &&
            this.main_complete()
        );
    },

    getSummary: function () {
        var skipSomatic = this.main_complete() && this.meets_criteria_none();
        return (
            taskcommon.formatDateSimple(this.date_pertains_to) + " " +
            this.get_main_description() + " " +
            (skipSomatic ? "" : this.get_somatic_description()) +
            this.isCompleteSuffix()
        );
    },

    detailgroup: function (arr) {
        var uifunc = require('lib/uifunc'),
            msg = "",
            i;
        for (i = 0; i < arr.length; ++i) {
            msg += taskcommon.descriptionValuePair(this, arr[i], arr[i],
                                                   uifunc.presentAbsentUnknown);
        }
        return msg;
    },

    getDetail: function () {
        var uifunc = require('lib/uifunc'),
            msg;
        msg = (
            taskcommon.descriptionValuePair(this, "date_pertains_to",
                                            "date_pertains_to",
                                            taskcommon.formatDateSimple) +
            taskcommon.descriptionValuePair(this, "examiners_comments",
                                            "comments") +
            taskcommon.descriptionValuePair(this, "duration_at_least_2_weeks",
                                            "duration_at_least_2_weeks",
                                            uifunc.trueFalseUnknown)
        );
        msg += this.detailgroup(CORE_NAMES) + "\n";
        msg += this.detailgroup(ADDITIONAL_NAMES) + "\n";
        msg += taskcommon.descriptionValuePair(this, "severe_clinically",
                                               "severe_clinically",
                                               uifunc.trueFalseUnknown) + "\n";
        msg += this.detailgroup(SOMATIC_NAMES) + "\n";
        msg += this.detailgroup(PSYCHOSIS_NAMES) + "\n";
        msg += this.getSummary();
        return msg;
    },

    edit: function (readOnly) {
        var self = this,
            Questionnaire = require('questionnaire/Questionnaire'),
            UICONSTANTS = require('common/UICONSTANTS'),
            true_false_options = taskcommon.OPTIONS_FALSE_TRUE_BOOLEAN,
            present_absent_options = taskcommon.OPTIONS_ABSENT_PRESENT_BOOLEAN,
            STEM = "icd10depressive_",
            pages,
            questionnaire;

        self.setDefaultClinicianVariablesAtFirstUse(readOnly); // Clinician info 2/3

        function editgroup(arr, mandatory) {
            return {
                type: "QuestionMCQGrid",
                horizontal: true,
                options: present_absent_options,
                questions: taskcommon.localizedStringArrayBySuffixArray(STEM,
                                                                        arr),
                fields: arr,
                mandatory: mandatory
            };
        }

        pages = [
            {
                title: L('t_icd10_depressive_episode'),
                clinician: true,
                elements: [
                    self.getClinicianQuestionnaireBlock(), // Clinician info 3/3
                    {
                        type: "QuestionText",
                        text: L('icd10_symptomatic_disclaimer')
                    },
                    {
                        type: "QuestionText",
                        text: L('date_pertains_to'),
                        bold: true
                    },
                    {
                        type: "QuestionDateTime",
                        field: "date_pertains_to",
                        showTime: false
                    },
                    {
                        type: "QuestionText",
                        text: L('icd10depressive_duration_text'),
                        bold: true
                    },
                    {
                        type: "QuestionMCQGrid",
                        horizontal: true,
                        options: true_false_options,
                        questions: [
                            L("icd10depressive_duration_at_least_2_weeks")
                        ],
                        fields: [
                            "duration_at_least_2_weeks"
                        ],
                        mandatory: true
                    },
                    {
                        type: "QuestionText",
                        text: L('icd10depressive_core'),
                        bold: true
                    },
                    editgroup(CORE_NAMES, true),
                    {
                        type: "QuestionText",
                        text: L('icd10depressive_additional'),
                        bold: true
                    },
                    editgroup(ADDITIONAL_NAMES, true),
                    {
                        type: "QuestionText",
                        text: L('icd10depressive_clinical_text'),
                        bold: true
                    },
                    {
                        type: "QuestionMCQGrid",
                        horizontal: true,
                        options: true_false_options,
                        questions: [
                            L("icd10depressive_severe_clinically")
                        ],
                        fields: [
                            "severe_clinically"
                        ],
                        mandatory: true
                    },
                    {
                        type: "QuestionText",
                        text: L('icd10depressive_somatic'),
                        bold: true
                    },
                    editgroup(SOMATIC_NAMES, false),
                    {
                        type: "QuestionText",
                        text: L('icd10depressive_psychotic'),
                        bold: true
                    },
                    editgroup(PSYCHOSIS_NAMES, true),
                    {
                        type: "QuestionTypedVariables",
                        mandatory: false,
                        useColumns: false,
                        variables: [
                            {
                                type: UICONSTANTS.TYPEDVAR_TEXT_MULTILINE,
                                field: "comments",
                                prompt: L("comments")
                            }
                        ]
                    }
                ]
            }
        ];

        questionnaire = new Questionnaire({
            readOnly: readOnly,
            pages: pages,
            callbackThis: self,
            fnGetFieldValue: self.defaultGetFieldValueFn,
            fnSetField: self.defaultSetFieldFn,
            fnFinished: self.defaultFinishedFn
        });
        questionnaire.open();
    }

});

module.exports = Icd10Depressive;

File: tablet_qt/tasks/icd10depressive.cpp
/*
    Copyright (C) 2012-2018 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.
*/

#define SHOW_CLASSIFICATION_WORKING

#include "icd10depressive.h"
#include "common/appstrings.h"
#include "common/textconst.h"
#include "lib/convert.h"
#include "lib/datetime.h"
#include "maths/mathfunc.h"
#include "lib/stringfunc.h"
#include "questionnairelib/commonoptions.h"
#include "questionnairelib/namevaluepair.h"
#include "questionnairelib/qudatetime.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/quheading.h"
#include "questionnairelib/qumcq.h"
#include "questionnairelib/qumcqgrid.h"
#include "questionnairelib/qutext.h"
#include "questionnairelib/qutextedit.h"
#include "tasklib/taskfactory.h"
using datetime::shortDate;
using mathfunc::anyNull;
using mathfunc::anyTrue;
using mathfunc::countNull;
using mathfunc::countTrue;
using stringfunc::bold;
using stringfunc::standardResult;

const QString Icd10Depressive::ICD10DEP_TABLENAME("icd10depressive");

const QString MOOD("mood");
const QString ANHEDONIA("anhedonia");
const QString ENERGY("energy");
const QStringList CORE_NAMES{
    MOOD,
    ANHEDONIA,
    ENERGY,
};
const QString SLEEP("sleep");
const QString WORTH("worth");
const QString APPETITE("appetite");
const QString GUILT("guilt");
const QString CONCENTRATION("concentration");
const QString ACTIVITY("activity");
const QString DEATH("death");
const QStringList ADDITIONAL_NAMES{
    SLEEP,
    WORTH,
    APPETITE,
    GUILT,
    CONCENTRATION,
    ACTIVITY,
    DEATH,
};
const QString SOMATIC_ANHEDONIA("somatic_anhedonia");
const QString SOMATIC_EMOTIONAL_UNREACTIVITY("somatic_emotional_unreactivity");
const QString SOMATIC_EARLY_MORNING_WAKING("somatic_early_morning_waking");
const QString SOMATIC_MOOD_WORSE_MORNING("somatic_mood_worse_morning");
const QString SOMATIC_PSYCHOMOTOR("somatic_psychomotor");
const QString SOMATIC_APPETITE("somatic_appetite");
const QString SOMATIC_WEIGHT("somatic_weight");
const QString SOMATIC_LIBIDO("somatic_libido");
const QStringList SOMATIC_NAMES{
    SOMATIC_ANHEDONIA,
    SOMATIC_EMOTIONAL_UNREACTIVITY,
    SOMATIC_EARLY_MORNING_WAKING,
    SOMATIC_MOOD_WORSE_MORNING,
    SOMATIC_PSYCHOMOTOR,
    SOMATIC_APPETITE,
    SOMATIC_WEIGHT,
    SOMATIC_LIBIDO,
};
const QString HALLUCINATIONS_SCHIZOPHRENIC("hallucinations_schizophrenic");
const QString HALLUCINATIONS_OTHER("hallucinations_other");
const QString DELUSIONS_SCHIZOPHRENIC("delusions_schizophrenic");
const QString DELUSIONS_OTHER("delusions_other");
const QString STUPOR("stupor");
const QStringList PSYCHOSIS_AND_SIMILAR_NAMES{
    HALLUCINATIONS_SCHIZOPHRENIC,
    HALLUCINATIONS_OTHER,
    DELUSIONS_SCHIZOPHRENIC,
    DELUSIONS_OTHER,
    STUPOR,
};
const QString DATE_PERTAINS_TO("date_pertains_to");
const QString COMMENTS("comments");
const QString DURATION_AT_LEAST_2_WEEKS("duration_at_least_2_weeks");
const QString SEVERE_CLINICALLY("severe_clinically");
const QStringList INFORMATIVE = CORE_NAMES + ADDITIONAL_NAMES +
        PSYCHOSIS_AND_SIMILAR_NAMES +
        QStringList{DURATION_AT_LEAST_2_WEEKS, SEVERE_CLINICALLY};


void initializeIcd10Depressive(TaskFactory& factory)
{
    static TaskRegistrar<Icd10Depressive> registered(factory);
}


Icd10Depressive::Icd10Depressive(CamcopsApp& app, DatabaseManager& db,
                                 const int load_pk) :
    Task(app, db, ICD10DEP_TABLENAME, false, true, false)  // ... anon, clin, resp
{
    addFields(CORE_NAMES, QVariant::Bool);
    addFields(ADDITIONAL_NAMES, QVariant::Bool);
    addFields(SOMATIC_NAMES, QVariant::Bool);
    addFields(PSYCHOSIS_AND_SIMILAR_NAMES, QVariant::Bool);

    addField(DATE_PERTAINS_TO, QVariant::Date);
    addField(COMMENTS, QVariant::String);
    addField(DURATION_AT_LEAST_2_WEEKS, QVariant::Bool);
    addField(SEVERE_CLINICALLY, QVariant::Bool);

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.

    // Extra initialization:
    if (load_pk == dbconst::NONEXISTENT_PK) {
        setValue(DATE_PERTAINS_TO, datetime::nowDate(), false);
    }
}


// ============================================================================
// Class info
// ============================================================================

QString Icd10Depressive::shortname() const
{
    return "ICD10-depression";
}


QString Icd10Depressive::longname() const
{
    return tr("ICD-10 symptomatic criteria for a depressive episode "
              "(as in e.g. F06.3, F25, F31, F32, F33)");
}


QString Icd10Depressive::menusubtitle() const
{
    return textconst::ICD10;
}


QString Icd10Depressive::infoFilenameStem() const
{
    return "icd";
}


// ============================================================================
// Instance info
// ============================================================================

bool Icd10Depressive::isComplete() const
{
    return !valueIsNull(DATE_PERTAINS_TO) && mainComplete();
}


QStringList Icd10Depressive::summary() const
{
    return QStringList{
        standardResult(appstring(appstrings::DATE_PERTAINS_TO),
                       shortDate(value(DATE_PERTAINS_TO))),
        standardResult(textconst::CATEGORY, getFullDescription()),
    };
}


QStringList Icd10Depressive::detail() const
{
    QStringList lines = completenessInfo();
    lines.append(standardResult(appstring(appstrings::DATE_PERTAINS_TO),
                                shortDate(value(DATE_PERTAINS_TO))));
    lines.append(fieldSummary(COMMENTS,
                              textconst::EXAMINER_COMMENTS));
    lines.append(fieldSummaryTrueFalseUnknown(DURATION_AT_LEAST_2_WEEKS,
                                              DURATION_AT_LEAST_2_WEEKS));
    lines.append("");
    lines += detailGroup(CORE_NAMES);
    lines += detailGroup(ADDITIONAL_NAMES);
    lines.append(fieldSummaryTrueFalseUnknown(SEVERE_CLINICALLY,
                                              SEVERE_CLINICALLY));
    lines += detailGroup(SOMATIC_NAMES);
    lines += detailGroup(PSYCHOSIS_AND_SIMILAR_NAMES);
    lines.append("");
    lines.append(standardResult(textconst::CATEGORY, getFullDescription()));
#ifdef SHOW_CLASSIFICATION_WORKING
    auto pv = [](const QVariant& v) -> QString {
        return bold(convert::prettyValue(v));
    };
    lines.append("");
    lines.append("nCore(): " + pv(nCore()));
    lines.append("nAdditional(): " + pv(nAdditional()));
    lines.append("nTotal(): " + pv(nTotal()));
    lines.append("nSomatic(): " + pv(nSomatic()));
    lines.append("mainComplete(): " + pv(mainComplete()));
    lines.append("meetsCriteriaSeverePsychoticSchizophrenic(): " +
                 pv(meetsCriteriaSeverePsychoticSchizophrenic()));
    lines.append("meetsCriteriaSeverePsychoticIcd(): " +
                 pv(meetsCriteriaSeverePsychoticIcd()));
    lines.append("meetsCriteriaSevereNonpsychotic(): " +
                 pv(meetsCriteriaSevereNonpsychotic()));
    lines.append("meetsCriteriaSevereIgnoringPsychosis(): " +
                 pv(meetsCriteriaSevereIgnoringPsychosis()));
    lines.append("meetsCriteriaModerate(): " + pv(meetsCriteriaModerate()));
    lines.append("meetsCriteriaMild(): " + pv(meetsCriteriaMild()));
    lines.append("meetsCriteriaNone(): " + pv(meetsCriteriaNone()));
    lines.append("meetsCriteriaSomatic(): " + pv(meetsCriteriaSomatic()));
#endif
    return lines;
}


OpenableWidget* Icd10Depressive::editor(const bool read_only)
{
    const NameValueOptions true_false_options = CommonOptions::falseTrueBoolean();
    const NameValueOptions present_absent_options = CommonOptions::absentPresentBoolean();

    auto heading = [this](const QString& xstringname) -> QuElement* {
        return new QuHeading(xstring(xstringname));
    };
    auto grid = [this](const QStringList& fields_xstrings,
                       const NameValueOptions& options,
                       bool mandatory) -> QuElement* {
        // Assumes the xstring name matches the fieldname (as it does)
        QVector<QuestionWithOneField> qfields;
        for (auto fieldname : fields_xstrings) {
            qfields.append(
                        QuestionWithOneField(xstring(fieldname),
                                             fieldRef(fieldname, mandatory)));
        }
        const int n = options.size();
        const QVector<int> v(n, 1);
        return (new QuMcqGrid(qfields, options))
                ->setExpand(true)
                ->setWidth(n, v);
    };

    QuPagePtr page((new QuPage{
        getClinicianQuestionnaireBlockRawPointer(),
        (new QuText(appstring(appstrings::ICD10_SYMPTOMATIC_DISCLAIMER)))->setBold(),
        new QuText(appstring(appstrings::DATE_PERTAINS_TO)),
        (new QuDateTime(fieldRef(DATE_PERTAINS_TO)))
            ->setMode(QuDateTime::Mode::DefaultDate)
            ->setOfferNowButton(true),
        heading("duration_text"),
        grid({DURATION_AT_LEAST_2_WEEKS}, true_false_options, true),
        heading("core"),
        grid(CORE_NAMES, present_absent_options, true),
        heading("additional"),
        grid(ADDITIONAL_NAMES, present_absent_options, true),
        grid({SEVERE_CLINICALLY}, true_false_options, true),
        heading("somatic"),
        grid(SOMATIC_NAMES, present_absent_options, false),
        heading("psychotic"),
        grid(PSYCHOSIS_AND_SIMILAR_NAMES, present_absent_options, false),
        new QuHeading(textconst::COMMENTS),
        new QuTextEdit(fieldRef(COMMENTS, false)),
    })->setTitle(longname()));

    for (auto fieldname : INFORMATIVE) {
        connect(fieldRef(fieldname).data(), &FieldRef::valueChanged,
                this, &Icd10Depressive::updateMandatory);
    }

    updateMandatory();

    Questionnaire* questionnaire = new Questionnaire(m_app, {page});
    questionnaire->setType(QuPage::PageType::Clinician);
    questionnaire->setReadOnly(read_only);
    return questionnaire;
}


// ============================================================================
// Task-specific calculations
// ============================================================================

int Icd10Depressive::nCore() const
{
    return countTrue(values(CORE_NAMES));
}


int Icd10Depressive::nAdditional() const
{
    return countTrue(values(ADDITIONAL_NAMES));
}


int Icd10Depressive::nTotal() const
{
    return nCore() + nAdditional();
}


int Icd10Depressive::nSomatic() const
{
    return countTrue(values(SOMATIC_NAMES));
}


bool Icd10Depressive::mainComplete() const
{
    if (valueBool(SEVERE_CLINICALLY)) {
        return true;  // depression present and definitely severe
    }
    if (valueIsFalseNotNull(DURATION_AT_LEAST_2_WEEKS)) {
        return true;  // depression absent; too short
    }

    const QVector<QVariant> core = values(CORE_NAMES);
    const int t_core = countTrue(core);
    const int u_core = countNull(core);
    if (t_core + u_core < 2) {
        return true;  // depression absent; definitely <2 core symptoms
    }

    const QVector<QVariant> additional = values(ADDITIONAL_NAMES);
    const int t_additional = countTrue(additional);
    const int u_additional = countNull(additional);

    if (t_core == 3 && (t_core + t_additional) >= 8) {
        return true;  // depression present and severe
    }
    if (t_core + u_core + t_additional + u_additional < 4) {
        return true;  // depression absent; <4 total symptoms
    }

    // If we get here: changes in the "unknown" bits might change presence/
    // absence, or severity.
    return false;
}


QVariant Icd10Depressive::meetsCriteriaSeverePsychoticSchizophrenic() const
{
    const QVariant severe = meetsCriteriaSevereIgnoringPsychosis();
    if (!severe.toBool()) {
        return severe;  // might be false or NULL
    }
    const QStringList icd10psychotic{STUPOR, HALLUCINATIONS_OTHER,
                                     DELUSIONS_OTHER};
    const QStringList schizophreniform{HALLUCINATIONS_SCHIZOPHRENIC,
                                       DELUSIONS_SCHIZOPHRENIC};
    if (anyTrue(values(icd10psychotic))) {
        return false;  // that counts as F32.3
    }
    if (anyNull(values(icd10psychotic))) {
        return QVariant();  // might be F32.3
    }
    if (anyTrue(values(schizophreniform))) {
        return true;
    }
    if (anyNull(values(schizophreniform))) {
        return QVariant();
    }
    return false;
}


QVariant Icd10Depressive::meetsCriteriaSeverePsychoticIcd() const
{
    const QVariant severe = meetsCriteriaSevereIgnoringPsychosis();
    if (!severe.toBool()) {
        return severe;  // might be false or NULL
    }
    // For psychotic depression (F32.3), the ICD-10 Green Book requires the
    // PRESENCE of non-schizophreniform psychotic symptoms, but not the ABSENCE
    // of schizophreniform psychotic symptoms.
    const QStringList icd10psychotic{STUPOR, HALLUCINATIONS_OTHER,
                                     DELUSIONS_OTHER};
    if (anyTrue(values(icd10psychotic))) {
        return true;
    }
    if (anyNull(values(icd10psychotic))) {
        return QVariant();
    }
    return false;
}


QVariant Icd10Depressive::meetsCriteriaSevereNonpsychotic() const
{
    const QVariant severe_ign_psy = meetsCriteriaSevereIgnoringPsychosis();
    if (!severe_ign_psy.toBool()) {
        return severe_ign_psy;  // might be false or NULL
    }
    if (anyNull(values(PSYCHOSIS_AND_SIMILAR_NAMES))) {
        return QVariant();
    }
    return countTrue(values(PSYCHOSIS_AND_SIMILAR_NAMES)) == 0;
}


QVariant Icd10Depressive::meetsCriteriaSevereIgnoringPsychosis() const
{
    if (valueBool(SEVERE_CLINICALLY)) {
        return true;
    }
    if (valueIsFalseNotNull(DURATION_AT_LEAST_2_WEEKS)) {
        return false;  // too short
    }
    if (nCore() >= 3 && nTotal() >= 8) {
        return true;  // ICD-10 definition of severe deperssion
    }
    if (!mainComplete()) {
        return QVariant();  // addition of more information might increase severity
    }
    return false;
}


QVariant Icd10Depressive::meetsCriteriaModerate() const
{
    if (meetsCriteriaSevereIgnoringPsychosis().toBool()) {
        return false;  // too severe
    }
    if (valueIsFalseNotNull(DURATION_AT_LEAST_2_WEEKS)) {
        return false;  // too short
    }
    if (!mainComplete()) {
        return QVariant();  // addition of more information might increase severity
    }
    if (nCore() >= 2 && nTotal() >= 6) {
        return true;  // ICD-10 definition of moderate depression
    }
    return false;
}


QVariant Icd10Depressive::meetsCriteriaMild() const
{
    if (meetsCriteriaSevereIgnoringPsychosis().toBool() ||
            meetsCriteriaModerate().toBool()) {
        return false;  // too severe
    }
    if (valueIsFalseNotNull(DURATION_AT_LEAST_2_WEEKS)) {
        return false;  // too short
    }
    if (!mainComplete()) {
        return QVariant();  // addition of more information might increase severity
    }
    if (nCore() >= 2 && nTotal() >= 4) {
        return true;  // ICD-10 definition of mild depression
    }
    return false;
}


QVariant Icd10Depressive::meetsCriteriaNone() const
{
    if (meetsCriteriaSevereIgnoringPsychosis().toBool() ||
            meetsCriteriaModerate().toBool() ||
            meetsCriteriaMild().toBool()) {
        return false;  // depression is present
    }
    if (valueIsFalseNotNull(DURATION_AT_LEAST_2_WEEKS)) {
        return true;  // too short to have depression
    }
    if (!mainComplete()) {
        return QVariant();  // addition of more information might increase severity
    }
    return true;
}


QVariant Icd10Depressive::meetsCriteriaSomatic() const
{
    const int t = nSomatic();  // t for true
    const int u = countNull(values(SOMATIC_NAMES));  // u for unknown
    if (t >= 4) {
        return true;
    }
    if (t + u < 4) {
        return false;
    }
    return QVariant();
}


QString Icd10Depressive::getSomaticDescription() const
{
    const QVariant s = meetsCriteriaSomatic();
    if (s.isNull()) {
        return xstring("category_somatic_unknown");
    }
    if (s.toBool()) {
        return xstring("category_with_somatic");
    }
    return xstring("category_without_somatic");
}


QString Icd10Depressive::getMainDescription() const
{
    if (meetsCriteriaSeverePsychoticSchizophrenic().toBool()) {
        return xstring("category_severe_psychotic_schizophrenic");
    }
    if (meetsCriteriaSeverePsychoticIcd().toBool()) {
        return xstring("category_severe_psychotic");
    }
    if (meetsCriteriaSevereNonpsychotic().toBool()) {
        return xstring("category_severe_nonpsychotic");
    }
    if (meetsCriteriaModerate().toBool()) {
        return xstring("category_moderate");
    }
    if (meetsCriteriaMild().toBool()) {
        return xstring("category_mild");
    }
    if (meetsCriteriaNone().toBool()) {
        return xstring("category_none");
    }
    return textconst::UNKNOWN;
}


QString Icd10Depressive::getFullDescription() const
{
    // I note in passing:
    //   QVariant v;
    //   bool b = bool(v);  // won't compile: invalid cast from type 'QVariant' to type 'bool'
    //   bool b = v;  // won't compile: cannot convert 'QVariant' to 'bool' in initialization
    //   bool b = true && v;  // won't compile: no match for 'operator&&' (operand types are 'bool' and 'QVariant')
    // ... which is good, as it means we can't forget the .toBool() part!
    const bool skip_somatic = mainComplete() && meetsCriteriaNone().toBool();
    return getMainDescription() + (skip_somatic
                                   ? ""
                                   : " " + getSomaticDescription());
}


QStringList Icd10Depressive::detailGroup(const QStringList& fieldnames) const
{
    QStringList lines;
    for (auto f : fieldnames) {
        lines.append(fieldSummary(f, f));
    }
    return lines;
}


// ============================================================================
// Signal handlers
// ============================================================================

void Icd10Depressive::updateMandatory()
{
    const bool known = meetsCriteriaNone().toBool() ||
            meetsCriteriaMild().toBool() ||
            meetsCriteriaModerate().toBool() ||
            meetsCriteriaSevereNonpsychotic().toBool() ||
            meetsCriteriaSeverePsychoticIcd().toBool() ||
            meetsCriteriaSeverePsychoticSchizophrenic().toBool();
    const bool need = !known;
    for (auto fieldname : INFORMATIVE) {
        fieldRef(fieldname)->setMandatory(need, this);
    }
}
